### PR TITLE
Hard soft excitation flag

### PR DESCRIPTION
**Pull request by RevathiJambunathan**
This PR adds the capability to either set the field value using the external excitation, called hard source
or add the external excitation on top of the self-consistent field, called soft source.
To do this, first a flag parser is used that spatially sets the type of excitation (hard=0/soft=1) for each field component
Then based on the value of the flag for a given (x,y,z), the external field is applied at that location.
The code will abort if the flag function is not specified for an external excitation read in using the parser.

I used the attached input file to test the implementation for USE_LLG=FALSE.

Verification test : 
Attached is an input file used to test the implementation.
An E-field excitation is added with corresponding flag types
```
my_constants.flag_none = -1
my_constants.flag_hs = 0
my_constants.flag_ss = 1
warpx.E_ext_grid_init_style = parse_E_ext_grid_function
warpx.Ex_external_grid_function(x,y,z) = "10.0"
warpx.Ey_external_grid_function(x,y,z) = "10.0"
warpx.Ez_external_grid_function(x,y,z) = "10.0"
warpx.E_excitation_on_grid_style = "parse_E_excitation_grid_function"
warpx.Ex_excitation_grid_function(x,y,z,t) = "0.0"
warpx.Ey_excitation_grid_function(x,y,z,t) = "100*(z<z0)+ 50*(z>=z0)"
warpx.Ez_excitation_grid_function(x,y,z,t) = "50.0"
warpx.Ex_excitation_flag_function(x,y,z) = "flag_none"
warpx.Ey_excitation_flag_function(x,y,z) = "flag_hs*(z<z0) + flag_ss*(z>=z0)"
warpx.Ez_excitation_flag_function(x,y,z) = "flag_ss"
```
I turned off the Maxwell solver in the code and initialized all the E-field components to have a value of 10 V/m and ran the simulation for two timesteps,

At the end of the two timesteps, ideally, Ex everywhere should be equal to 10 V/m and the plot below confirms that.
Screenshot from 2020-10-20 16-44-09

Ez should be 10 + 50 + 50 = 110 V/m everywhere.
Screenshot from 2020-10-20 16-45-44

Finally for Ey : for z < 0, the value should be strictly 100 and z>=0, the value should be 10 + 50 + 50 = 110.
At z==0 because the plotfiles are cell-centered and it take the average of the adjacent edge-based Ey values, the value is 105.
(This is only for the diagnostic and not the raw data)
Screenshot from 2020-10-20 16-48-18

[inputs_3d_test_flag.txt](https://github.com/RevathiJambunathan/WarpX/files/5412234/inputs_3d_test_flag.txt)

### CODE REVIEW

File: Source/WarpX.cpp [hunk 1]
```diff
@@ -86,6 +86,13 @@ std::string WarpX::str_Hx_excitation_grid_function;
 std::string WarpX::str_Hy_excitation_grid_function;
 std::string WarpX::str_Hz_excitation_grid_function;
 #endif
+// Flag for type of excitation (hard/soft)
```
Comment (ajnonaka) [inline Source/WarpX.cpp]:
```suggestion
// Flag for type of excitation -1=none, 0=soft, 1=hard
```

File: Source/WarpX.cpp [hunk 2]
```diff
@@ -86,6 +86,13 @@ std::string WarpX::str_Hx_excitation_grid_function;
...
+std::string WarpX::str_Ex_excitation_flag_function = "0";
+std::string WarpX::str_Ey_excitation_flag_function = "0";
+std::string WarpX::str_Ez_excitation_flag_function = "0";
+std::string WarpX::str_Bx_excitation_flag_function = "0";
+std::string WarpX::str_By_excitation_flag_function = "0";
+std::string WarpX::str_Bz_excitation_flag_function = "0";
```
Comment (ajnonaka) [inline Source/WarpX.cpp]: Consider making the default value "-1" so the source is off unless a user explicitly states where the source(s) should be.

**Review by ajnonaka — CHANGES_REQUESTED** [review]
Note that we also have to update the documentation for B_excitation_on_grid_style to also indicate that you need to specify the Bx_excitation_flag_function, etc.  Otherwise it looks good.

File: Source/FieldSolver/WarpXExternalEMFields.cpp [hunk 3]
```diff
@@ -85,19 +94,34 @@ WarpX::ApplyExternalFieldExcitationOnGrid (
                 amrex::Real x, y, z;
                 WarpXUtilAlgo::getCellCoordinates(i, j, k, mfx_stag,
                                                   problo, dx, x, y, z);
-                Fx(i, j, k) += xfield_parser(x, y, z, t);
+                if ( xflag_parser(x,y,z) == 0 ) {
+                    Fx(i, j, k) += xfield_parser(x, y, z, t);
+                }
+                if ( xflag_parser(x,y,z) == 1 ) {
+                    Fx(i, j, k) = xfield_parser(x, y, z, t);
+                }
```
Comment (jackieyao0114) [inline Source/FieldSolver/WarpXExternalEMFields.cpp]: Would it look okay to change this to:
`Fx(i, j, k) = Fx(i, j, k) * xflag_parser(x,y,z) + xfield_parser(x, y, z, t);`

Comment (RevathiJambunathan) [reply Source/FieldSolver/WarpXExternalEMFields.cpp]: @jackieyao0114 Thank you for this solution! I think it is a great idea provided the flag_parser is 0/1.
But now we have a default of -1 to force the user to specify the type of excitation.

Comment (ajnonaka) [reply Source/FieldSolver/WarpXExternalEMFields.cpp]: 0 would have to mean hard source, and 1 would have to be soft source.
And you would only enter this is flag_parser(i,j,k) >= 0
I'm ok with this as long as a comment is left in the code explaining this.

Comment (ajnonaka) [reply Source/FieldSolver/WarpXExternalEMFields.cpp]: (hard is =, and soft is +=, right?)

Comment (ajnonaka) [reply Source/FieldSolver/WarpXExternalEMFields.cpp]: it's also a clarity thing... code is easier to understand the way it is
If the code is left as is, you may as well change the second "if" statement to and "else if" for a tiny speedup.

Comment (jackieyao0114) [reply Source/FieldSolver/WarpXExternalEMFields.cpp]: yeah, hard is = and soft is +=. So if we use my suggested code we have to flip the flags. I'm okay either way --- keep it as is or change it

Comment (RevathiJambunathan) [reply Source/FieldSolver/WarpXExternalEMFields.cpp]: just changed it to else if

File: Source/FieldSolver/WarpXExternalEMFields.cpp [hunk 4]
```diff
@@ -85,19 +94,37 @@ WarpX::ApplyExternalFieldExcitationOnGrid (
                 amrex::Real x, y, z;
                 WarpXUtilAlgo::getCellCoordinates(i, j, k, mfx_stag,
                                                   problo, dx, x, y, z);
-                Fx(i, j, k) += xfield_parser(x, y, z, t);
+                auto tmp_field_value = xfield_parser(x,y,z,t);
```
Comment (ajnonaka) [inline Source/FieldSolver/WarpXExternalEMFields.cpp]: Maybe we don't ALWAYS want to call the parser.  For most runs when flag_type=-1, this is just wasting CPU time.

// 0 is hard source, 1 is soft source - update comment below
if (flag_type >= 0) {
   Fx = flag_type * Fx + parsed_value
}

File: Source/FieldSolver/WarpXExternalEMFields.cpp (hunk 3 above)
Comment (RevathiJambunathan) [reply Source/FieldSolver/WarpXExternalEMFields.cpp]: But I will be happy to add this one-line elegance as well. and change the definition of hard-source and soft source in the flag! :) (within if (flag>=0) ...)

File: Source/FieldSolver/WarpXExternalEMFields.cpp (hunk 4 above)
Comment (jackieyao0114) [reply Source/FieldSolver/WarpXExternalEMFields.cpp]: Yes!

File: Source/WarpX.cpp (hunk 2 above)
Comment (RevathiJambunathan) [reply Source/WarpX.cpp]: done. thank you

**Review by ajnonaka — APPROVED** [review]
Approve pending the switch of flag_type to 0/1/2 for nothing/hard/soft
Approve pending error checking to make sure flag_type is 0.0 1.0 or 2.0

**Pull request merged**